specification/communication/data-plane/JobRouter/preview/2021-10-20-preview2/communicationservicejobrouter.json

```diff
@@ -0,0 +1,3635 @@
+        "jobId",
+        "workerId"
+      ],
+      "type": "object",
+      "properties": {
+        "assignmentId": {
+          "description": "The assignment Id that assigns a worker that has accepted an offer to a job.",
+          "type": "string"
+        },
+        "jobId": {
+          "description": "The Id of the job assigned.",
+          "type": "string"
+        },
+        "workerId": {
+          "description": "The Id of the worker that has been assigned this job.",
+          "type": "string"
+        }
+      }
+    },
+    "AzureFunctionRule": {
+      "description": "A rule providing a binding to an HTTP Triggered Azure Function.",
+      "required": [
+        "functionUrl"
+      ],
+      "type": "object",
+      "allOf": [
+        {
+          "$ref": "#/definitions/RouterRule"
+        }
+      ],
+      "properties": {
+        "kind": {
+          "type": "string",
+          "readOnly": true
+        },
+        "functionUrl": {
+          "format": "uri",
+          "description": "URL for Azure Function",
+          "type": "string"
+        },
+        "credential": {
+          "$ref": "#/definitions/AzureFunctionRuleCredential"
+        }
+      },
+      "x-ms-discriminator-value": "azure-function-rule"
+    },
+    "AzureFunctionRuleCredential": {
+      "description": "Credentials used to access Azure function rule",
+      "type": "object",
+      "properties": {
+        "functionKey": {
+          "description": "(Optional) Access key scoped to a particular function",
+          "maxLength": 500,
+          "minLength": 0,
+          "type": "string"
+        },
+        "appKey": {
+          "description": "(Optional) Access key scoped to a Azure Function app.\r\nThis key grants access to all functions under the app.",
+          "maxLength": 500,
+          "minLength": 0,
+          "type": "string"
+        },
+        "clientId": {
+          "description": "(Optional) Client id, when AppKey is provided\r\nIn context of Azure function, this is usually the name of the key",
+          "maxLength": 500,
+          "minLength": 0,
+          "type": "string"
+        }
+      }
+    },
+    "BestWorkerMode": {
+      "description": "Jobs are distributed to the worker with the strongest abilities available.",
+      "type": "object",
+      "allOf": [
+        {
+          "$ref": "#/definitions/DistributionMode"
+        }
+      ],
+      "properties": {
+        "kind": {
+          "type": "string",
+          "readOnly": true
+        },
+        "scoringRule": {
+          "description": "(Optional)\r\nDefine a scoring rule to use, when calculating a score to determine the best worker.\r\nIf not set, will use a default scoring formula that uses the number of job labels that the worker\r\nlabels match, as well as the number of label selectors the worker labels match and/or exceed\r\nusing a logistic function (https://en.wikipedia.org/wiki/Logistic_function).",
+          "x-nullable": true
```

Comment:
You should avoid the use of `x-nullable: true`. Properties with no value should simply be omitted from the payload.

Reply:
Tracked using [issue](https://github.com/Azure/azure-rest-api-specs/pull/17872)

Reply:
After investigation it seems like it's beneficial for client sdks to keep `x-nullable: true`. 

When we removed x-nullable flag, the client was modified from `property = null` to property.ThrowNonNullablePropertyIsNull(). This will throw JsonException in customer applications in Debug build.

"This exception only happens in DEBUG builds of the library and would be ignored in the release build" -- https://cs.github.com/Azure/autorest.csharp/blob/6915dbe22302dd54ea7883d0c51bdb917c0adb6c/src/assets/Generator.Shared/JsonElementExtensions.cs#L95

More discussion is here: https://skype.visualstudio.com/SCC/_git/ic3_spool_ccaap/pullrequest/603633?discussionId=6011870

Reply:
I've opened [this issue](https://github.com/Azure/autorest.csharp/issues/2068) and requested a quick resolution from the autorest team.